log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -0,0 +1,579 @@
+                    rollFile();
+                }
+                catch (IOException e) {
+                    // it is possible the roll worked, but there was a problem cleaning up, and it is possible it failed
+                    // either way, roll will be attempted again once file grows by maxFileSize
+                    currentFileSize = 0;
+
+                    reportError("Error rolling log file", e, GENERIC_FAILURE);
+                }
+            }
+
+            logHistoryManager.pruneLogFilesIfNecessary(currentFileSize + message.length);
+
+            currentFileSize += message.length;
+            try {
+                currentOutputStream.write(message);
+            }
+            catch (Exception e) {
+                reportError(null, e, WRITE_FAILURE);
+            }
+        }
+        // always flush at the end of a batch, so logs aren't delayed
+        flush();
+    }
+
+    private synchronized void rollFile()
+            throws IOException
+    {
+        // carefully update the stream, so if there is a problem logging can continue
+
+        LogFileName newFileName = null;
+        Path newFile = null;
+        OutputStream newOutputStream = null;
+        for (int i = 0; i < 100; i++) {
+            try {
+                newFileName = LogFileName.newLogFileName(link, compressionType.getExtension());
+                newFile = link.resolveSibling(newFileName.getFileName());
+                newOutputStream = new BufferedOutputStream(Files.newOutputStream(newFile, CREATE_NEW), MAX_BATCH_BYTES);
+                break;
+            }
+            catch (FileAlreadyExistsException ignore) {
+            }
+        }
+
+        // If a new file can not be opened, abort and continue using the existing file
+        if (newOutputStream == null) {
+            throw new IOException("Could not create new a unique log file");
+        }
+
+        // The new file is open, so we will always switch to this new output stream
+        // If any error occurs, with the cleanup steps, we add them to this exception as suppressed and throw at the end
+        IOException exception = new IOException(String.format("Unable to %s log file", currentOutputStream == null ? "setup initial" : "roll"));
+
+        if (currentOutputStream != null) {
+            try {
+                currentOutputStream.close();
+            }
+            catch (IOException e) {
+                exception.addSuppressed(new IOException("Unable to close old output stream", e));
+            }
+            logHistoryManager.addFile(currentOutputFile, currentOutputFileName, currentFileSize);
+            if (compressionExecutor != null) {
+                Path originalFile = currentOutputFile;
+                LogFileName originalLogFileName = currentOutputFileName;
+                long originalFileSize = currentFileSize;
+                compressionExecutor.submit(() -> compressInternal(originalFile, originalLogFileName, originalFileSize));
```

Comment:
It's probably ok for now, but I think it'd be cleaner and easier to follow the code if this was done unconditionally. The `compressInternal` method would skip if compression is not needed.

BTW, the fact that this method (`rollFile`) is being called in the constructor before `compressionExecutor` is set, and that the behavior of this method depends on whether `currentOutputStream` is set (which doesn't happen on the first run) makes it hard to reason about the interaction between all the parts. I don't have better ideas right now -- hopefully will develop some as I get a better picture of how everything fits together -- but you may want to think about whether that could be simplified.

Reply:
in the case of the constructor the `currentOutputStream` will be `null`. I added comment explaining that.

Reply:
I agree, but I don't want the unused compressor when we are not compressing.  Initially, I tried to put the compression in a separate manager, but the concurrency of this code makes adding that abstraction nearly impossible.